### ISSUE TITLE
Proposal: Manage CRDs

### ISSUE THREAD

**Issue by mattfarina** [helm/helm#5871]
I have to give credit to @prydonius for this idea. I'm just writing it up here. Can do a proposal to community if this needs to move forward. Not entirely sure what is feasible yet.

**The idea:**

Have a `crd` directory alongside the `templates` directory. Helm would then check for missing CRDs and attempt to install them. If it cannot install them and they are missing it would stop, explain why it can't install them (e.g., permissions), and provide some details (e.g., someone with permissions needs to install the CRDs).

This is because there is common wrapper logic needed around CRDs for each of them that is about the same for this. First, check if it's installed and if it's not then have it there with the crd install hook.

One unknown is handing changes in the CRD content at the same `apiVersion`. This happens in practice and has many potential problems. Can we detect a change and throw an error to require human intervention?

Once this is in place the Helm dependency system can be used to handle CRDs that are dependent upon one another. Each can be in their own chart with their own controllers.

**Comment by mattfarina** [comment]
cc @technosophos @adamreese @bacongobbler @hickeyma @michelleN @viglesiasce @thomastaylor312 @SlickNik

**Comment by bacongobbler** [comment]
I agree that there is a common wrapper logic being used for all CRDs in Helm 2 recently.

I would love to see a practical example! It seems like it's an idea worth investigating. I'm having a hard time conceptualizing this, so having a practical example to play with and understand the solution might be a good way forward to see if this is feasible or not (as well as provide feedback on the design).

**Comment by technosophos** [comment]
As I see it, we have four major problems to sort out with CRDs:

- Install order: CRDs have to be installed in a separate operation from any _usage of_ (instances of) that CRD. 
- Compatibility detection: If a CRD exists, and a request to install the CRD comes in, but the api versions are different, what do we do?
- Update process: CRDs are possibly global, while most other kinds that appear in Helm charts are not. So under what conditions can a chart that contains a CRD upgrade the CRD?
- Deletion of a CRD: When a CRD is deleted, all of its instances are deleted. And CRDs are global. Under what conditions can Helm delete a CRD (clearly, it's not by default)

As far as I can tell, the above makes it easier to solve the first point. But is it in any way helping or hindering with the others? I don't want to take a shot at solving just one prong, only to discover that that solution prevented us from solving the others.

The proposal does not explain what the mechanics of the `crd/` directory are. Are CRD YAML files treated as templates, too? Do they get the same context as other templates? Are they always evaluated BEFORE the others? And if so, can anything be injected from the CRD template eval into the context used for regular eval? What happens if a non-CRD is placed in the `crd` directory? For example, if I have a deployment that is a CRD controller, it is often necessary for that controller to be up and running before instances can be reliably processed. So do I put the controller objects in the CRD folder?

**Comment by mattfarina** [comment]
Just [spitballing](https://www.merriam-webster.com/dictionary/spitball)...

> Install order: CRDs have to be installed in a separate operation from any usage of (instances of) that CRD.

The CRD directory would make it easy to figure out where they were to install them first. We would not need to parse all the templates first. This could extend to the whole dependency tree.

> Compatibility detection: If a CRD exists, and a request to install the CRD comes in, but the api versions are different, what do we do?

This is the hard one. [There is a whole page in the docs on versioning CRDs](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/). While I'm open to suggestions, I would argue we notify the user and let them decide what to do.

> Update process: CRDs are possibly global, while most other kinds that appear in Helm charts are not. So under what conditions can a chart that contains a CRD upgrade the CRD?

When there is a change such as a new apiVersion, change in storage/serving, etc. I would propose this is an operation that notifies the user of the coming change and asks them to accept it. A human in the loop because this could be a destructive change.

> Deletion of a CRD: When a CRD is deleted, all of its instances are deleted. And CRDs are global. Under what conditions can Helm delete a CRD (clearly, it's not by default)

Always manual? Directions about deleting them with big flashers like we used to have on geocities for the under construction pages.

@technosophos you raised a lot of good questions. Ones worth airing and discussing publicly.

**Comment by prydonius** [comment]
> The CRD directory would make it easy to figure out where they were to install them first. We would not need to parse all the templates first. This could extend to the whole dependency tree.

Placing CRDs in a specific directory in the chart package gives us a more reliable way to handle them separately. For example, we could render all the CRDs in the chart and sub-charts first and then install them before installing the rest (just an example, the idea is we can use it to make the right decisions). I guess there's no reason we can't do this with the current `crd-install` hook, but I'm not sure what the future of hooks are in Helm 3 and I also feel like annotations aren't the best way of defining these if we have the opportunity to extend the chart format.

Completely agree with @mattfarina's comment, Helm should not try to do anything clever in most of these scenarios, I think the goal here would be to just provide more info to the operator so that they can decide how to proceed.

**Comment by paulczar** [comment]
`Have a crd directory alongside the templates directory`   - is this necessary?  We know that a CRD has the following top level items in their declaration:

```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
```

would it be within Helm's scope to look at the apiVersion/kind of resources and act on them based on the resource type?  If we know its a CRD by the apiVersion/kind then we can treat it as a CRD, however we decide to treat them.

**Comment by mattfarina** [comment]
@paulczar If Helm is to read the `apiVersion` and `kind` of each object and know how to act on them it's a new level of introspection and ordering.

So, let's say you have a CRD that needs to be installed, then a controller to work on those CRDs, and then some CRs you have a definite ordering thing. You have to know a fair amount of intent. Do we need to tell people to break these into two charts?

Then there is the `apiVersion`. What we have now is a beta version and that will change. At the least is to go to GA. How do we handle this changing in the future?

**Comment by prydonius** [comment]
Well FWIW Helm already defines the operation order per kind (https://github.com/helm/helm/blob/bed4054c412f95d140c8c98b6387f40df7f3139e/pkg/tiller/kind_sorter.go), so perhaps extracting the CRDs is not a terrible approach.

Not sure that the apiVersion matters that much, at least for core resources we can pretty much assume the kind name won't change or clash with anything (at least from beta onwards).

**Comment by paulczar** [comment]
So putting `crd` at the top of the sort order would be a good start.  Maybe helm can then check if the crd already exists with the `name` and `apiVersion`, and if so skip it, if not install it.   We could potentially do a equivalent of `kubectl auth can-i create crds` to validate the user has permissions to install crds, and have a friendly error.

In a larger picture we could also check each resource kind/apiVersion in the chart against `kubectl apiVersion`, `kubectl apiResources`, `kubectl get crd` to validate that its doable on the cluster (maybe a set of `kubectl auth can-i` to validate the user is allowed to ?), if it doesn't it can error out.

**Comment by bacongobbler** [comment]
> So putting crd at the top of the sort order would be a good start.

One of the problems with CRDs today stems from the fact that the CRD and its children (the CRs) are defined within the same chart. During resource validation, the CRs fail API validation because the CRD does not exist yet therefore there's no API endpoint that can accept those CRs. #2994 and @mattfarina's earlier comment summarizes the issue fairly well here.

Throwing CRDs earlier in the sort order will not help in this situation. They need to be installed prior to validating CRs within the chart itself... Either that or tell people that CRDs and CRs need to be broken into two charts.

However, that does bring up a good point. What if we validated and installed each resource in sort order, rather than only installation?

We might be able to change the current logic in pkg/kube from

1. render all templates as resources
2. validate all resources together <-- where CRDs/CRs currently fail unless they were installed using the `crd-install` hook
3. install resources based on sort order

To

1. render all templates as resources
2. based on sort order, for each GroupVersion
  a. validate the resources
  b. install all resources with the same GroupVersion
  c. GOTO next GroupVersion

It may be more computationally expensive, but it could solve the install order issue. Perhaps there may be issues with validating resources individually... @technosophos do you know of any limitations this approach may bring?

If that works, that would solve the "install order" issue and remove the need for a `crd` directory. That doesn't solve compatibility detection, update process or the CRD deletion use cases though.

> Then there is the `apiVersion`. What we have now is a beta version and that will change. At the least is to go to GA. How do we handle this changing in the future?

Did we have a v1beta2 to apps/v1 migration story for the workloads API? As far as I understand, handling apiVersion migrations is out of scope for kubernetes as the behaviour between different apiVersions may differ. That being said, there have been iterations in the past where very few changes occurred (like v1beta1 to v1beta2). I'm not sure how (or if) we can handle that.

**Comment by mattfarina** [comment]
Detecting if a CRD is installed and installing if not present is the easy problem to solve or bike shed on. Can we look at the hard problems instead? The things that don't have easy solutions.

For example, how do we handle updating CRDs? For example, A CRD is installed and there's an operator using it in one namespace. Now, someone is installing a second operator for the same thing in a different namespace without knowledge of the first install. The CRD for the second chart has new version information (maybe even a new storage version). Does Helm install the update? What will happen to the first operator? I think this has the potential to cause problems if I read https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects correctly.

Or, How do we handle charts that have CRDs, controllers, and CRs based on the CRDs. In this case the install order needs to be the CRD, the controller responding to the CRD, and then any other assets. If the controller is run as a deployment than it needs to be sort ordered after the CRD and before any other assets. How do we determine that?

The Create step in the CRD CRUD is the easiest to solve. It's the RUD parts that are more complicated and we need proposed solutions for.

**Comment by bacongobbler** [comment]
> Can we look at the hard problems instead?

Absolutely. Thank you for calling us out on that.

> For example, A CRD is installed and there's an operator using it in one namespace. Now, someone is installing a second operator for the same thing in a different namespace without knowledge of the first install. The CRD for the second chart has new version information (maybe even a new storage version). Does Helm install the update?

Ooh, good one. Yeah, this is a semi-common use case today where two applications rely on the same CRD. It's a good case of sharing common resources between multiple apps: how do the two reconcile a shared resource without stepping on eachother's toes?

If I'm not mistaken, this particular problem applies to all global resources including CRDs. For example, it's quite possible that two apps could create the same PersistentVolume or a similar StorageClass.

I guess the same problem might also be considered for non-globally scoped resources as well. I've heard users ask about how two applications installed in the same namespace can rely on a shared postgres instance in the same namespace. There might be similar issues there too.

Crazy thought here: what if Helm *didn't* manage globally-scoped resources, and left that to the cluster administrator? That way we leave the reconciliation of globally scoped resources to the cluster administrator.

Other than that, I don't have any immediate solutions that come to mind. How do other package managers reconcile shared resources between two apps? Do they just tell package managers that it's out of scope?

**Comment by paulczar** [comment]
> It's a good case of sharing common resources between multiple apps: how do the two reconcile a shared resource without stepping on eachother's toes?

If we take a stance of A helm chart being a standalone package, then we probably shouldn't try and handle two apps sharing a resource, instead we would make it three apps.

In this case where the shared resource is a CRD, then the CRD would be the third app, and would have its own method of install ( whether it being a helm chart that's installed first, or some other method ).

That would make our stance something like "you can use helm to manage a CRD, but it should be its own discreet package, and not be part of an operator or other chart".   The user could then reach for a higher order tool like `helmfile` to orchestrate composing infrastructure from multiple charts.

> Crazy thought here: what if Helm didn't manage globally-scoped resources, and left that to the cluster administrator? That way we leave the reconciliation of globally scoped resources to the cluster administrator.

I'm starting to lean in this direct.  A CRD is effectively adding to the kubernetes control plane, an app deployed by a "user" definitely shouldn't be trying to change the behaviour of the control plane, and the blast radius from a CRD CRUD operation is big enough that Helm probably shouldn't try and manage the lifecycle of them.  Hell its a hard enough problem that the operator-sdk folks have built a tool specifically to deal with operator lifecycle management - https://github.com/operator-framework/operator-lifecycle-manager

**Comment by bacongobbler** [comment]
From Slack, courtesy of @jzelinskie:

> This is a rabbithole that OLM has tried to manage for years. I believe there's a working group where we're trying to refactor/redesign something that replaces the kube addon manager that handles lifecycles of CRDs

> I don't think any decisions in Helm should be made without consulting that wg

**Comment by mattfarina** [comment]
@bacongobbler Thanks for chasing that down. Two thoughts...

While I'm glad people are thinking through the addon elements to CRDs, [Brendan documented multiple use cases for these in a PR](https://github.com/kubernetes/enhancements/pull/833) of which addons was only one of the cases.

How does this all work for the operator pattern? Take for example, the etcd operator. What if it is installed into more than one namespace by more than one team. How do we handle that case? It's not really scoped the same as a cluster addon.

**Comment by jzelinskie** [comment]
>How does this all work for the operator pattern? Take for example, the etcd operator. What if it is installed into more than one namespace by more than one team. How do we handle that case? It's not really scoped the same as a cluster addon.

This is something that has been flip-flopped on a bunch in the past development of OLM. At first, we tried to follow the target of least privilege. This meant that namespaces would each get their own instances of operators but OLM would manage conflicts with installing CRDs at the cluster-scope. Eventually, we were finding that many people cared about cluster-scoped operators, which have their own problems (they don't always scale to huge clusters, RBAC can be tricky, etc...). To satisfy both, OLM introduced yet another resource type OperatorGroups. This basically lets you define the set of namespaces for a single "tenant" of operators. We spent time in the multi-tenancy work group, but honestly, Kubernetes was not designed for real multi-tenancy and it's going to be REALLY hard if not impossible to add retroactively. OpenShift.io is probably the only real multi-tenant k8s that exists and it's a special snowflake that probably no one else should emulate.

**Comment by hickeyma** [comment]
From @joejulian in #6065:

I had a user complain that he lost all his records that were associated with a CRD when a release was deleted. We changed his chart and set the crd-install hook. Later, this same user complained that he couldn't delete and reinstall the same chart because it failed to create the CRD (it was already there). Deleting the CRD via kubectl was not an option as it would lose all the records associated with the CRD and the user was uncomfortable with using kubectl as he was only familiar with using helm. Further, he wanted multiple releases of his software in other namespaces which, again, caused an error because the CRD was already installed.

In trying to preserve that user experience and, at the same time, that user's data we handled that by adding a value that could be changed to install the CRD (or not). This caused confusion because multiple users may install this chart into their namespace, but one has to install that chart with the CRD installation flag set. Which one? How do they know in advance?

Again, this was a less than happy user experience.

- the user would like to install a chart
- if there is no CRD, they would like it created
- if there is a CRD, they would like the rest of the chart to be installed
- they don't want to have to know in advance to set a flag to disable the creation of the CRD
- they don't want to lose their data when a release is deleted. This data may be assocated with other releases.

Suggestion - since CRDs can only be installed once and don't hold state, I propose:

- if a chart has a CRD defined, and that CRD is identical to the one installed, the installation of that object is skipped and the rest of the objects are installed as normal
- if the CRD differs from the one installed, and no flags are set on install, the installation would fail stating that the CRD is installed and differs from the one in the chart.
- if running helm upgrade with or without the install flag, if the CRD differs from the one installed tiller attempts to upgrade the CRD
- if running helm upgrade with or without the install flag, if the CRD is identical to the one installed the installation of the CRD is skipped

**Comment by mattfarina** [comment]
Currently, Helm v2 does not handle CRDs. We are working to fix this before beta that will carry a new way to handle CRDs in Helm v3. Here is the proposal we are working on...

1) CRDs will go in a new directory in a chart. Helm will attempt to install CRDs if they are not present in the cluster before rendering templates or generating the details for Capabilities within templates. In the Helm v2 capabilities are generated only once meaning that once CRDs are added they do not appear in the Capabilities object for other resources.

The two step process is because the installation of CRDs can't really be fully declarative. You have CRDs, controllers that act on them, and CRs leveraging those that need to be a bit ordered. Here we will provide the ability to provide CRDs in a manner that is obviously ordered and separate.

2) The `crd-install` hook will not work in Helm v3. The `apiVersion` of `v2` for charts will not support it. For those working with `apiVersion` of `v1` for charts (the Helm v2 charts) there will be plugin that can be installed for Helm v3 that will provide support. The details of how the plugin will work are still in development.

CRD workflows are not fully baked within the community.  Attempting to encode some experimental ideas of workflows in Helm is something we'd like to avoid. Hopefully we can encode some processes once they are baked in the community. To be baked I'd hope they are well documented, practiced, and generally agreed to.

I have some ideas that are not fully part of the plan. Maybe something for experimentation with a plugin to help move the ball forward in the future.

If there are questions or comments on this please feel free to speak up. There is quite a bit of nuance to the whole process and some subtle issues when you go down different paths. Supporting the `crd-install` hook in Helm v3 would be a significant amount of work at this point and make the code a lot more complex. Codifying processes that aren't backed in the community could be bad if those processes don't take off.

/cc @adamreese @thomastaylor312 @technosophos

**Comment by thomastaylor312** [comment]
See #6243 for an implementation

**Comment by hickeyma** [comment]
@mattfarina Does the merge of #6243 implement this proposal?

**Comment by TBBle** [comment]
Did either of the plugins mentioned by @mattfarina in https://github.com/helm/helm/pull/6243#pullrequestreview-276609696 ever see the light of day? Since this ticket is still open, is it the right place to discuss such, or are there other forums where CRD behaviours other than "install once and forever" are still being worked through?

**Comment by thomastaylor312** [comment]
@TBBle We eventually decided that the legacy plugin was not needed as you can have both Helm 2 style CRD install hooks and Helm 3 crds/ directory in the same chart. The other mentioned plugin was just a thought, but not something actively being worked on (if you have ideas or code you could share, they'd be much appreciated!). 

But yes, this is the proper place to discuss CRD things. I think there is also discussions on CRDs in general in SIG-Apps (I think?).

**Comment by TBBle** [comment]
I haven't gone digging through Slack, the mailing list, or listened to the meeting recordings, but I checked the [SIG Apps Meeting Minutes](https://docs.google.com/document/d/1LZLBGW2wRDwAfdBNHJjFfk9CFoyZPcIYGWU7R1PQ3ng/edit#) via https://github.com/kubernetes/community/tree/master/sig-apps.

The only things that seemed relevant came from @mattfarina anyway:

> [*July 1st 2019*](https://youtu.be/svMSUg2ARDo?t=69) (Discussion is 9 minutes):
> CRD management - when an operator is installed more than once, two namespaces that do not relate to each other. How do we manage that? [Matt Farina]
> * work with API machinery to document some of this

> [*Feb 4th 2019*](https://youtu.be/vLXwlWE-Ajg?t=2760) (Discussion is 14 minutes):
> CRD version lifecycle management workflows (mattfarina)
> * Deletion is not really an option right now. Will delete all artifacts using them
> * Now is a good time to get involved. Specifically with API machinery
> * * SDK sub-project is trying to come up with machinery on this
> * Not a great idea to update a CRD at same version
> * * Adding a feature should work
> * * Changing validation in a way that doesn’t break any existing validations should work
> * * Needs to be strict backwards compatibility
> * * Protocol buffers does something where they make a copy of a field and deprecate the old one
> * CRDs are cluster wide and don’t work for all orgs

If I take those as starting points, as well the primary use-case that affects me ("CRDs are immutable" on https://github.com/googleforgames/agones/issues/1280#issuecomment-575286428), then I'd suggest this as a possible direction, leaning upon CRD multi-versioning support and schemas.

1. Helm 3 processing _crds/_ in a chart, only warns if the CRDs are *different*. (haven't checked, perhaps this is already the case)
2. CRDs placed in _crds/_ must be backwards-compatible with previous releases. So when an existing version is modified:
   * An OpenAPI 3 schema is required (already mandatory in `apiextensions.k8s.io/v1`. but was optional in `apiextensions.k8s.io/v1beta1`)
   * New fields are optional/defaulted
   * Fields are not removed nor change type, unless moving to a less-restrictive format
   * e.g., larger integer range, removing something like `format: uuid`
   * I think this means they have to be ["structural"](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#specifying-a-structural-schema), for the above to be reliable?
   * Any other changes are handled by a new [CRD version and migration strategy](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/).
      *The iterative changes that are my focus here would mostly not need this, or be fine with the 'None' migration which prunes deleted fields.
      * Narrowing integer fields or changing types is the biggest risk.
      * This requires k8s v1.13 for automatic storage conversion, and even then ["webhook" conversion](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion) was only alpha, so I don't know how widely it was deployed.
3. A new Helm plugin is written, e.g., `update-crds`, which can take a chart, look at its CRDs, and decide if the CRD in the chart is newer than the one on the cluster, and update if appropriate.
   * Unless we want to make it a requirement of the CRDs, the plugin must understand how to [compare a single-version v1beta1 CRD with a multi-version v1beta1 or v1 CRD](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresource-conversion-webhook.md#top-level-fields-to-per-version-fields).
   * A CRD with a strict superset of versions (or whose highest-priority version is higher, perhaps) is "newer".
      * Dropping an older version in a new CRD is not unreasonable, but are there risks here?
   * Otherwise, for the latest version in the CRD, the strict superset of fields wins.
   * At this point, we're looking at field types. I'd be *thrilled* if there's a library out there that can tell us when a OpenAPI 3 schema is a strict backwards-compatible superset of another schema.
  * Failure to determine an order means a warning and no action taken.
  * There's always a `--force` option (or `--recreate`?) which just overwrites the CRDs even if they appear to be older, or are not safely orderable, or do not meet the other criteria, e.g. schema-less CRDS.
  * This plugin would also make it easy to give a Helm chart to your cluster-admin and say
    > Please install the CRDs from this chart, so I the non-privileged namespace-limited person can run `helm install --skip-crds` without further interrupting to [your busy day](https://sysadminday.com/).
4. Helm 3's warning could point at the plugin, which *should* be safe according to the above to run first anyway, as existing custom resources will remain valid.
5. If the plugin's behaviour proves sound and widely-useful, eventually it could migrate into Helm's CRD workflow, to replace the 'equality' check, and instead skip-and-warn if a CRD appears to be newer on the cluster than in the Chart.
   * `helm install/upgrade --crd-only` would be a nice complement to `helm install --skip-crds`. And since they're not templated, this should be very simple for users to use.

It's definitely reasonable that a cluster has a newer CRD definition than a chart being installed, since CRDs are cluster-wide, but different application versions may be present at the same time, particularly third-party sub-charts.

Disclaimer: I'm only really learning a lot of these systems now, and I'm coming at them in the place of a user. So I may have faulty assumptions here, and welcome correction/redirection.

**Comment by munnerz** [comment]
So it doesn’t get lost in the noise, x-ref https://github.com/helm/helm/issues/7735 as there was quite a bit of discussion there as to what chart maintainers that are releasing CRDs are after, and various members of sig-api-machinery would expect from deployment tools 🙂

FWIW, in cert-manager at least we have opted to not use the ‘crds’ sub directory at all, nor do we use the ‘crd-install’ annotation (as per all the discussion here!). Instead we just include CRDs as part of our template/ directory, which gives us full management/control over these resources.

This works fine so long as nobody declares our chart as a dependency of their own, although we have long advised against this as conceptually our CRDs/controllers should be thought of as ‘control plane components’ and should be installed separate to other apps 🙂

Personally, I’d add a huge -1 for Helm doing anything special to handle the lifecycle of CRDs. If two charts both declare the same CRD, this should be classed as an error. It may on the surface seem sensible to try making an intelligent decision about which one of two resources to use, or how to merge them, but in practice could most definitely lead to unintended and hard to understand side effects. CRDs should use a vendor suffix, and no single vendor should publish two charts that contains the same CRD.

All I really want is for Helm to do as little as possible to special case CRD resources, aside from what the behaviour of the ‘crd-install’ originally intended to do (installing an instance of a CR within the same chart as the CRD that defines its type).

**Comment by Morriz** [comment]
Yes, please don't become catch all guardians of possibly-broken-charts. Let's promote proper testing strategies to guard ourselves against broken software, instead of forcing CRDs to be static (which they are obviously not, just like APIs).

**Comment by joejulian** [comment]
FWIW, we (D2iQ) had to fork the cert-manager chart (and a bunch of others) _because_ of how it handles CRDs. Istio has abandoned helm and developed their own installer binary because of CRD management.

I agree with the majority of this topic that CRD management is a problem that needs to be addressed.

**Comment by technosophos** [comment]
The problem is simple, and if anyone has a solution, write a proposal and then a PR.

The problem: CRDs are global resources. If a CRD is modified in an incompatible way, it can break all usages of that CRD in the cluster. CRDs are in charts, and it is unreasonable to assume that an operator will read every line of change in a chart. Evidence suggests that it is also unreasonable to assume that chart authors will care enough about installations to prevent compatibility breaks. The consequence of this is that it is likely that updating CRDs will introduce substantial breaking behavior that will impact users beyond just the ones who used the chart.

If anyone can figure out a way to fix this, then please propose it.

But simply saying, "this is the user's problem" doesn't fly. We're the ones who will have to field the issues filed by angry people who accidentally destroy their cluster integrity with a bad upgrade. We're the ones who get at-replied on angry tweet storms. And we're the ones who have to respond to public criticism of the project when someone's chart doesn't do versioning correctly. So please don't tell us that this isn't a valid reason. You're not the ones who will have to deal with the headache.

**Comment by mattfarina** [comment]
@joejulian you are absolutely right that CRD handing is a problem that needs to be addressed. There is a little context that might help and some things you might be able to help with.

First, the context...

- There is no direct guidance from the Kubernetes project on handling of CRDs.
- Those who have attempted to handle CRDs themselves only deal with a subset of situations while ignoring other valid situations. They are not complete solutions.
- An example of that, I can have two distinct things running in a cluster that are not aware of each other but both rely on the same CRD. How do you handle things like upgrades of the CRD and who owns the CRD to approve those changes?

Second, how you can help...

- Do you have a proposed solution you could share?
- Do you have specific use cases to share that the solution needs to fit?

The problem we've run into with proposals so far is that they only solve part of the problem for some people. They are not complete solutions or near complete solutions. Common CRD situations are skipped. On Helm we have to consider more than one path.

Also, we work very hard to not break users. Move fast and break things isn't what Helm is about because it's a building block to other things. Imagine if apt or yum moved fast and broke things.

**Comment by Morriz** [comment]
> The problem is simple, and if anyone has a solution, write a proposal and then a PR.
> 
> 
> 
> The problem: CRDs are global resources. If a CRD is modified in an incompatible way, it can break all usages of that CRD in the cluster. CRDs are in charts, and it is unreasonable to assume that an operator will read every line of change in a chart. Evidence suggests that it is also unreasonable to assume that chart authors will care enough about installations to prevent compatibility breaks. The consequence of this is that it is likely that updating CRDs will introduce substantial breaking behavior that will impact users beyond just the ones who used the chart.
> 
> 
> 
> If anyone can figure out a way to fix this, then please propose it.
> 
> 
> 
> But simply saying, "this is the user's problem" doesn't fly. We're the ones who will have to field the issues filed by angry people who accidentally destroy their cluster integrity with a bad upgrade. We're the ones who get at-replied on angry tweet storms. And we're the ones who have to respond to public criticism of the project when someone's chart doesn't do versioning correctly. So please don't tell us that this isn't a valid reason. You're not the ones who will have to deal with the headache.

I hear you. It’s not fun to deal with an angry mob that expects nobody to break working software (even themselves). But that problem is not the issue here. If you want to deal with the expectancies of a user base then I suggest you look at other parties, because many are doing that successfully.


But that we have entered an era of dependency complexities does still not warrant a fear driven approach that enforces version freezes. All other mature package managers have solved that problem, and imo they still exist because they did so elegantly.

I don’t see the above reasoning as sustainable, deviating from best practices when it comes to versioning. Please don’t reinvent this wheel and support developers doing proper versioning. I kindly ask to follow the paradigm that works and has been working for decades, instead of breaking it. (pun intended)

**Comment by bacongobbler** [comment]
As @technosophos and @mattfarina both mentioned previously, we are looking for concrete use cases on how Helm should handle CRD upgrades. While your idea sounds enticing, it doesn't explain the "how" or the "why" this approach would work for the use cases described above. If you can share a more concrete solution, that would be ideal.

I ask because I cannot comprehend what you are describing here. Do you have a practical example that demonstrates your idea, and how that solves the use cases described earlier in this thread?

**Comment by bacongobbler** [comment]
More specifically, I'm asking for more details on this and how it applies to Helm:

> I kindly ask to follow the paradigm that works and has been working for decades

**Comment by monotek** [comment]
> The problem: CRDs are global resources. If a CRD is modified in an incompatible way, it can break all usages of that CRD in the cluster. CRDs are in charts, and it is unreasonable to assume that an operator will read every line of change in a chart. Evidence suggests that it is also unreasonable to assume that chart authors will care enough about installations to prevent compatibility breaks. The consequence of this is that it is likely that updating CRDs will introduce substantial breaking behavior that will impact users beyond just the ones who used the chart.
> 
> If anyone can figure out a way to fix this, then please propose it.
> 
> But simply saying, "this is the user's problem" doesn't fly. We're the ones who will have to field the issues filed by angry people who accidentally destroy their cluster integrity with a bad upgrade. We're the ones who get at-replied on angry tweet storms. And we're the ones who have to respond to public criticism of the project when someone's chart doesn't do versioning correctly. So please don't tell us that this isn't a valid reason. You're not the ones who will have to deal with the headache.

In my opinion correct semver usage and information about upgrade process should be responsibility of the chart maintainer and therefore not tried to be prevented by helm itself. 

This prevents not only wrong usage but usage at all.

If helm ist "The package manager for Kubernetes" it should be possible to update packages whith all kind of resources. 

If you compare Helm to APT or RPM such packages could destroy your system too if the package contains some broken rm -rf for example but imho nobody would blame the package manager for not catching such broken things in the install scripts.

**Comment by stuart-c** [comment]
There are however big differences between Helm and Apt/RPM in how they operate. Those package managers are only for the use of a single user (root) and details of all packages are stored in a central database. Also, each package maintains a complex list of dependencies between packages and versions: requirements, conflicts, etc.

If you compare it to Helm there are a number of differences. There is no central database of what is installed, indeed different users could be totally isolated from each other. It is perfectly valid for users to be restricted to their own namespace and have no idea about what is installed elsewhere on the cluster, and if Helm (or any other or no) management system is used.

As a result some of the things you depend upon to not break your system don't exist. If you try to remove libc RPM will automatically uninstall anything that depends on it. If you try to upgrade a library APT will stop you if there are other packages which wouldn't work with the new version.

How do you manage that sort of check, if you in your namespace don't know what charts are installed elsewhere (and shouldn't be able to know as the isolation between namespaces is there to hide customers/teams/departments from each other).

**Comment by moduspwnens** [comment]
The concern is more unique to cluster-scoped resources than CRDs specifically, right?

If **Chart B** depends on a ClusterRole defined in **Chart A**, and an update to **Chart A** updates that ClusterRole to remove some permission **Chart B** depends upon, the net result is the same: the **Chart B** release is now broken due to revisions made when updating the **Chart A** release.

The unique (and critical) aspect of CRDs is that a deletion could result in more data loss than is expected (although correct me if I'm wrong here). But truly any non-backward compatible change to a cluster-scoped resource could break things for other releases that depend on it.

I guess I'm curious if this could be simplified with a clearer delineation between cluster-scoped and namespace-scoped resource types. The former are likely more comparable to apt or yum: system-wide changes requiring heavy permissions / user confirmation, and the latter would cover the less risky "just in my own namespace" kind of installs.

Like if when you try to use helm to upgrade / delete a release that would result in at least one cluster-scoped resource being deleted, it'd give you a big warning explaining that it might break things in other releases that depend on it. And perhaps a scarier warning for CRDs that includes the number of instances of that resource that'll go away. It might avoid some frustrated users opening tickets if they had to type **YES** after being told it'll delete their CRs.

Although it might also need a flag for just skipping the add / upgrade / delete of cluster-scoped resources to fit the use case targeted with the "crds" directory. 

But I recognize I'm probably forgetting or misunderstanding others' use cases, as this is a complex issue. And I do appreciate the maintainers' openness and deliberate approach here.

**Comment by stuart-c** [comment]
Yes other cluster wide resources can be problematic too, but how many charts would depend on something like a ClusterRole as opposed to a CRD?

Being able to not deploy cluster wide resources would be important for the sue case where the user only has permissions in their own namespace, linked with the opposite case where someone wants to only install the CRDs (admin user asked to do so to allow restricted users to work).

The trouble with giving a user a warning are twofold. Firstly how do we even know about the other usages? I might have permissions to my namespace and globally to CRDs, but not to look within other namespaces. Secondly if the user is informed of other charts which might break what are the options? Helm couldn't upgrade those other charts (like APT/RPM would), so presumably we are back to either stopping or going ahead regardless.

**Comment by Morriz** [comment]
Let's reiterate. Version management for k8s CRDs is coming in 1.16. Can we work on that and keep whatever helm does now for versions below?

https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/

**Comment by moduspwnens** [comment]
@stuart-c 
> Yes other cluster wide resources can be problematic too, but how many charts would depend on something like a ClusterRole as opposed to a CRD?

ClusterRoles are pretty commonly used in charts.
https://github.com/helm/charts/search?q=ClusterRole&unscoped_q=ClusterRole

But I may be misunderstanding your point. I think ClusterRoles are less problematic because they are easier for the chart maintainer to upgrade in a backward-compatible way than CRDs, and there's no risk of unexpected data loss if it's deleted and replaced. Regardless, it's still a cluster-scoped resource that could be modified in a way that's not backward-compatible and break other installs.

> Being able to not deploy cluster wide resources would be important for the sue case where the user only has permissions in their own namespace, linked with the opposite case where someone wants to only install the CRDs (admin user asked to do so to allow restricted users to work).

I'd expect the admin to want to install all cluster-scoped resources in that case, right? Not just the CRDs? If s/he's OK with the user installing ClusterRoles but not CRDs, then s/he's just rolling the dice on whether or not a chart update to ClusterRoles breaks others' installs.

> The trouble with giving a user a warning are twofold. Firstly how do we even know about the other usages? I might have permissions to my namespace and globally to CRDs, but not to look within other namespaces.

Helm doesn't have to know about the other usages. The warning tells the user that there may be other usages, and forces them to acknowledge it.

I mentioned potentially showing a number of CR instances that would be deleted, and that'd require permissions the user might not have, but that's just for a clearer warning. You can still warn them without that number if you don't have permission.

> Secondly if the user is informed of other charts which might break what are the options? Helm couldn't upgrade those other charts (like APT/RPM would), so presumably we are back to either stopping or going ahead regardless.

We are, but now the user knows what they're about to do. S/he isn't deleting a chart without realizing it could impact other installs (or cause CR instances to be deleted, if a CRD is being deleted).

**Comment by TBBle** [comment]
The difference between ClusterRoles  (and other cluster-wide resources) and CRDS is that ClusterRoles can have a different name that doesn't affect their behaviour because the things that reference them can also use that different name.

CRDs are special because they are _designed_ to be the interaction point between disconnected systems, so they need to exist at their known names, or they are not useful.

Also, deleting CRDs deletes _other things_. No other cluster resources do that directly; they may do it by action of an operator or controller, of course.

To contrast with rpm and apt, as has been done earlier, a CRD is like _/var/lib/mariadb_. Sure, it was created by a package, but if you delete it when you remove the package, you delete the user data too.

If two users define a ClusterRole differently through Helm values, which should win? Wider, narrower, union, intersection, first, last?

Unlike other cluster-wide resources, CRDs have an evolution mechanism and multi-versioning system defined [since 1.11](https://github.com/kubernetes/kubernetes/blob/release-1.11/CHANGELOG-1.11.md#sig-api-machinery).

For CRDs, we know when we can resolve this safely (one will be backwards-compatible to the other) and otherwise we need the administrator to make a choice. Ideally everyone's Helm charts will only ship backwards-compatible CRDs, and so rarely will we need the administrator to be explicit.


For me, the right conceptual behaviour is that Helm only ever silently installs or upgrades CRDs, and it supports explicitly downgrading or force-replacing, and it _never_ deletes them.

----

Helm's current approach is to not touch CRDs, except that it will install them for you if you don't have them already. Otherwise it's up to the cluster administrators, i.e. the ones who have permission to install CRDs and other Cluster-wide resources, to maintain those CRDs, similarly to how they would maintain them without Helm.

This is a _starting point_, it's not "mission accomplished" for CRDs. Per https://github.com/helm/helm/issues/7735#issuecomment-617795653, the lowest-hanging fruit could be an "--overwrite-crds" flag to complement the "--skip-crds" flag which lets you explicitly both trust that your chart producer didn't break their CRDs between releases, and that _you_ are installing the latest release on the cluster, and aren't going to break some other admin's installation of a newer version of the same operator.

My [mooted proposal](https://github.com/helm/helm/issues/5871#issuecomment-581145634) earlier is intended to go further than that, and make the known-safe cases silently correct, and unsafe cases explicit as to why they are unsafe, and allow a resolution.

It's worth including in this discussion that there are operators like https://github.com/jaegertracing/jaeger-operator which are designed to support multi-install, each operator monitoring only their own namespace, but which are going to need to share CRDs. So we can't assume the [cert-manager policy](https://github.com/helm/helm/issues/5871#issuecomment-624462551) of "CRDs are just global resources, and you'll only ever have one cert-manager instance installed on your cluster" applies to all Helm charts.

**Comment by lucendio** [comment]
Hello everyone,

I tried to go through all the discussions to see if the following thought has bee brought up already. I couldn't find any evidence, but I may have overlooked something. If that's the case, please point me into the right direction. Anyway, I'm looking forward to your feedback.

AFAIU, [CRDs allow the definition and availability of multiple API versions at the same time](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#overview), which are identified by a `name` (`spec.versions.*.name`). These `apiVersion`s are referred to in every instance of the corresponding custom resource.

So, every time Helm tries to apply a CRD, It could check whether a API version (`spec.versions.*.name`) either even existed in the previous release version, or if the `schema` of a specific API version is still identical to the schema in the previous release version. If it's different, Helm throws. This way, CRD authors can only introduce (breaking) changes by introducing a new API version entry in `spec.versions`.

**Comment by kyounger** [comment]
Would considering an option like `--only-crds` make sense, where helm would skip everything _except_ the crds? One of the two recommended approaches in the docs is to install CRDs as a separate chart. This would allow reuse of the chart package for two release installs, 1) as a release for the CRDs at a specific version, and 2) a release of the chart templates at that same version.

I think this would also work well with k8s api versioning. You would just install multiple releases of the chart (using proposed `--only-crds`) at different versions and you'd have the ability to take advantage of multiple CRD versions.

I readily admit that it doesn't trigger my "elegance" spidey-sense. But typically these CRDs are developed in the same git repo and versioned together anyway. So I think we all agree it makes sense to keep them in the same chart. This would just provide a way to keep that pattern and easily implement recommend docs approach #2.

**Comment by TBBle** [comment]
That would probably make more sense as a command, `install-crds`, I think. Since they're not templated, for example, a lot of the options (like `--values`) wouldn't make much sense to have available. It'd be specifically-useful in the situation I mentioned above where you want to use a chart that depends on CRDs, but do not yourself have permissions to install CRDs on a cluster, so you can ask the administrator to _just_ install the CRDs for you.

I still intend to try and implement my ideas above (both `--force-crds`, `--overwrite-crds`, or similar as the inversion of `--skip-crds`, and a (plugin, at first) crd-management command (I called it `update-crds` earlier, it probably overlaps with the `install-crds` idea) which can attempt to correctly update CRDs by default, and when it cannot, gives the user the right levers to do it, with an eye to building that into Helm's upgrade/install flows.

**Comment by monotek** [comment]
How about that workflow:

If "helm upgrade" detects that a crd would be overwritten, it should fail with something like: "crd has same version. Use --force to overwrite it anyway or --skip-crd to leave it untouched."

If a crd with a new version is found it should be installed additionally, keeping the old version in place too.

**Comment by bacongobbler** [comment]
See https://github.com/helm/community/pull/138 for the explanation why we are not considering that approach.

**Comment by KlavsKlavsen** [comment]
If I understand this correctly, Helm project has abandoned trying to support CRD lifecycle.. which means "something else" will have to be written to update CRDs in Helm charts?

**Comment by monotek** [comment]
FluxCD can ;-)

https://fluxcd.io/docs/components/helm/helmreleases/#crds

**Comment by kvaps** [comment]
Heya, I just added some workaround plugin to solve this problem:
https://github.com/kvaps/helm-ensurecrds
